File: Java/JavaSE/OpenGL/source/core/src/loon/action/sprite/effect/PShadowEffect.java
package loon.action.sprite.effect;

import loon.action.sprite.ISprite;
import loon.core.LObject;
import loon.core.LSystem;
import loon.core.geom.RectBox;
import loon.core.graphics.LColor;
import loon.core.graphics.LImage;
import loon.core.graphics.opengl.GLEx;
import loon.core.graphics.opengl.LTexture;
import loon.core.timer.LTimer;
import loon.jni.NativeSupport;
import loon.utils.CollectionUtils;


/**
 * Copyright 2008 - 2011
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 * 
 * @project loonframework
 * @author chenpeng
 * @email：ceponline@yahoo.com.cn
 * @version 0.1
 */
// 这是一个根据导入的图片黑白象素分布来完成渐变效果的特殊类，根据导入的渐变图不同，能够衍生出无穷多的渐变效果。
// 此种方式在吉里吉里(krkr)，nscript等AVG游戏引擎中较常使用。（因此，也可以直接套用它们的渐变图）
public class PShadowEffect extends LObject implements ISprite {

	private static final long serialVersionUID = 1L;

	private LTimer timer = new LTimer(100);

	private int x, y, width, height, scaleWidth, scaleHeight, pixelCount,
			layer;

	static int[] deasilTrans, widdershinTrans;

	private int[] nowDrawPixels, finalDrawPixels;

	private int[] backgroundPixels, finalBackgroundPixels;

	private boolean visible = true, flag = true, isDirty, isClose;

	private LTexture texture;

	private LImage image;

	private final static int max_pixel = 256;

	private final static int min_pixel = 0;

	private int pixelSkip = 8;

	public PShadowEffect(String fileName) {
		this(LImage.createImage(fileName));
	}

	public PShadowEffect(LImage img) {
		this(img, 0, 0);
	}

	public PShadowEffect(String fileName, String backFile) {
		this(LImage.createImage(fileName), LImage.createImage(backFile), 0, 0,
				LSystem.screenRect.width, LSystem.screenRect.height);
	}

	public PShadowEffect(LImage img, int x, int y) {
		this(img, null, x, y, img.getWidth(), img.getHeight());
	}

	public PShadowEffect(String fileName, int x, int y, int w, int h) {
		this(LImage.createImage(fileName), null, x, y, w, h);
	}

	public PShadowEffect(LImage img, LImage back, int x, int y) {
		this(img, back, x, y, img.getWidth(), img.getHeight());
	}

	public PShadowEffect(String fileName, String bacFile, int x, int y, int w,
			int h) {
		this(LImage.createImage(fileName), LImage.createImage(bacFile), x, y,
				w, h);
	}

	public PShadowEffect(String fileName, LImage back, int x, int y, int w,
			int h) {
		this(LImage.createImage(fileName), back, x, y, w, h);
	}

	private PixelThread pixelThread;

	public PShadowEffect(LImage img, LImage back, int x, int y, int w, int h) {
		if (deasilTrans == null || widdershinTrans == null) {
			deasilTrans = new int[max_pixel];
			for (int i = 0; i < max_pixel; i++) {
				deasilTrans[i] = LColor.getRGB(i, i, i);
			}
			int flag = 0;
			widdershinTrans = new int[max_pixel];
			for (int i = 0; i < max_pixel; i++) {
				widdershinTrans[flag++] = deasilTrans[i];
			}
		}
		this.x = x;
		this.y = y;
		this.width = w;
		this.height = h;
		this.visible = true;
		LImage temp = null;
		// 此处与Android版处理有差异，因为微机上无论如何不会慢……
		this.scaleWidth = width;
		this.scaleHeight = height;
		if (back == null) {
			temp = img.scaledInstance(scaleWidth, scaleHeight);
			this.texture = new LTexture(scaleWidth, scaleHeight, true);
			this.image = new LImage(scaleWidth, scaleHeight, true);
			this.finalDrawPixels = temp.getPixels();
			this.nowDrawPixels = CollectionUtils.copyOf(finalDrawPixels);
			if (temp != null) {
				temp.dispose();
				temp = null;
			}
		} else {
			temp = img.scaledInstance(scaleWidth, scaleHeight);
			this.texture = new LTexture(scaleWidth, scaleHeight, true);
			this.image = new LImage(scaleWidth, scaleHeight, true);
			if (back.getWidth() == scaleWidth
					&& back.getHeight() == scaleHeight) {
				this.finalBackgroundPixels = back.getPixels();
				this.backgroundPixels = CollectionUtils
						.copyOf(finalBackgroundPixels);
			} else {
				LImage tmp = back.scaledInstance(scaleWidth, scaleHeight);
				this.finalBackgroundPixels = tmp.getPixels();
				if (tmp != null) {
					tmp.dispose();
					tmp = null;
				}
				this.backgroundPixels = CollectionUtils
						.copyOf(finalBackgroundPixels);
			}
			this.finalDrawPixels = temp.getPixels();
			this.nowDrawPixels = CollectionUtils.copyOf(finalDrawPixels);
		}
		this.setBlackToWhite(flag);
		if (temp != null) {
			temp.dispose();
			temp = null;
		}
		if (img != null) {
			img.dispose();
			img = null;
		}
		if (back != null) {
			back.dispose();
			back = null;
		}
	}

	public void reset() {
		if (isClose) {
			return;
		}
		if (flag) {
			pixelCount = min_pixel;
		} else {
			pixelCount = max_pixel;
		}
		this.visible = true;
		this.nowDrawPixels = CollectionUtils.copyOf(finalDrawPixels);
		this.backgroundPixels = CollectionUtils.copyOf(finalBackgroundPixels);
		this.startUsePixelThread();
	}

	private final static int BLACK = LColor.black.getRGB();

	private final static int WHITE = LColor.white.getRGB();

	public void createUI(GLEx g) {
		if (isClose) {
			return;
		}
		if (!visible) {
			return;
		}
		synchronized (texture) {
			if (alpha > 0 && alpha < 1) {
				g.setAlpha(alpha);
			}
			if (!isComplete() && isDirty) {
				g.copyImageToTexture(texture, image, 0, 0);
				g.drawTexture(texture, x, y, width, height);
				isDirty = false;
			} else if (!isComplete()) {
				g.drawTexture(texture, x, y, width, height);
			}
			if (alpha > 0 && alpha < 1) {
				g.setAlpha(1f);
			}
		}
	}

	private long elapsed;

	public void update(long elapsedTime) {
		this.elapsed = elapsedTime;
	}

	private class PixelThread extends Thread {
		public void run() {
			for (; !isClose && !isComplete();) {
				if (image == null) {
					return;
				}
				if (visible && timer.action(elapsed)) {
					if (backgroundPixels == null) {
						if (flag) {
							NativeSupport.filterColor(max_pixel, pixelCount,
									pixelCount += pixelSkip, finalDrawPixels,
									nowDrawPixels, widdershinTrans, BLACK,
									WHITE);
						} else {
							NativeSupport.filterColor(max_pixel, pixelCount,
									pixelCount -= pixelSkip, finalDrawPixels,
									nowDrawPixels, deasilTrans, BLACK, WHITE);
						}
						image.setPixels(nowDrawPixels, scaleWidth, scaleHeight);
					} else {
						if (flag) {
							NativeSupport.filterColor(max_pixel, pixelCount,
									pixelCount += pixelSkip, finalDrawPixels,
									backgroundPixels, widdershinTrans, BLACK,
									WHITE);
						} else {
							NativeSupport
									.filterColor(max_pixel, pixelCount,
											pixelCount -= pixelSkip,
											finalDrawPixels, backgroundPixels,
											deasilTrans, BLACK, WHITE);
						}
						image.setPixels(backgroundPixels, scaleWidth,
								scaleHeight);
					}
					isDirty = true;
				}
			}
		}
	}

	public boolean isVisible() {
		return visible;
	}

	public void setVisible(boolean visible) {
		this.visible = visible;
	}

	final void startUsePixelThread() {
		if (pixelThread == null) {
			pixelThread = new PixelThread();
			pixelThread.start();
		}
	}

	final void endUsePixelThread() {
		if (pixelThread != null) {
			try {
				pixelThread.interrupt();
				pixelThread = null;
			} catch (Exception ex) {
				pixelThread = null;
			}
		}
	}

	public boolean isComplete() {
		final boolean stop = flag ? (pixelCount > max_pixel)
				: (pixelCount < min_pixel);
		if (!stop) {
			startUsePixelThread();
		} else {
			endUsePixelThread();
		}
		return stop;
	}

	public void setDelay(long delay) {
		timer.setDelay(delay);
	}

	public long getDelay() {
		return timer.getDelay();
	}

	public int getHeight() {
		return height;
	}

	public int getWidth() {
		return width;
	}

	public boolean isBlackToWhite() {
		return flag;
	}

	public void setBlackToWhite(boolean flag) {
		this.flag = flag;
		if (flag) {
			pixelCount = min_pixel;
		} else {
			pixelCount = max_pixel;
		}
	}

	public int getLayer() {
		return layer;
	}

	public void setLayer(int layer) {
		this.layer = layer;
	}

	public float getX() {
		return x;
	}

	public float getY() {
		return y;
	}

	public RectBox getCollisionBox() {
		return getRect(x, y, width, height);
	}

	public int x() {
		return x;
	}

	public int y() {
		return y;
	}

	public LTexture getBitmap() {
		return texture;
	}

	public boolean isClose() {
		return isClose;
	}

	public int getPixelCount() {
		return pixelCount;
	}

	public int getPixelSkip() {
		return pixelSkip;
	}

	public void setPixelSkip(int pixelSkip) {
		this.pixelSkip = pixelSkip;
	}

	public void dispose() {
		this.isClose = true;
		this.endUsePixelThread();
		this.finalDrawPixels = null;
		this.nowDrawPixels = null;
		if (texture != null) {
			texture.destroy();
			texture = null;
		}
		if (image != null) {
			image.dispose();
			image = null;
		}
	}

}

File: C#/WindowsPhone/LGame-XNA-lib/Loon.Action.Sprite.Effect/PShadowEffect.cs
using System;
using Microsoft.Xna.Framework;
using Loon.Core;
using Loon.Core.Timer;
using Loon.Core.Graphics.Opengl;
using Loon.Utils;
using Loon.Core.Graphics;
using Loon.Core.Geom;
using Loon.Java;

namespace Loon.Action.Sprite.Effect
{
    public class PShadowEffect : LObject, ISprite
    {

        private LTimer timer = new LTimer(100);

        private int width, height, scaleWidth, scaleHeight, pixelCount;

        static Color[] deasilTrans, widdershinTrans;

        private Color[] nowDrawPixels, finalDrawPixels;

        private Color[] backgroundPixels, finalBackgroundPixels;

        private bool visible = true, flag = true, isDirty, isClose;

        private LImage image;

        private static int max_pixel = 256;

        private static int min_pixel = 0;

        private int pixelSkip = 8;

        public PShadowEffect(string fileName)
            : this(LImage.CreateImage(fileName))
        {

        }

        public PShadowEffect(LImage img)
            : this(img, 0, 0)
        {

        }

        public PShadowEffect(string fileName, string backFile)
            : this(LImage.CreateImage(fileName), LImage.CreateImage(backFile), 0, 0,
                LSystem.screenRect.width, LSystem.screenRect.height)
        {

        }

        public PShadowEffect(LImage img, int x, int y)
            : this(img, null, x, y, img.GetWidth(), img.GetHeight())
        {

        }

        public PShadowEffect(string fileName, int x, int y, int w, int h)
            : this(LImage.CreateImage(fileName), null, x, y, w,
                h)
        {

        }

        public PShadowEffect(LImage img, LImage back, int x, int y)
            : this(img, back, x, y, img.GetWidth(), img.GetHeight())
        {

        }

        public PShadowEffect(string fileName, string bacFile, int x, int y, int w,
                int h)
            : this(LImage.CreateImage(fileName), LImage.CreateImage(bacFile), x, y, w, h)
        {

        }

        public PShadowEffect(string fileName, LImage back, int x, int y, int w,
                int h)
            : this(LImage.CreateImage(fileName), back, x, y, w,
                h)
        {

        }

        private PixelThread pixelThread;

        public PShadowEffect(LImage img, LImage back, int x, int y, int w, int h)
        {
            if (deasilTrans == null || widdershinTrans == null)
            {
                deasilTrans = new Color[max_pixel];
                for (int i = 0; i < max_pixel; i++)
                {
                    deasilTrans[i] = new Color(i, i, i);
                }
                int count = 0;
                widdershinTrans = new Color[max_pixel];
                for (int i = 0; i < max_pixel; i++)
                {
                    widdershinTrans[count++] = deasilTrans[i];
                }
            }
            this.SetLocation(x, y);
            this.width = w;
            this.height = h;
            this.visible = true;
            LImage temp = null;
            if (back == null)
            {
                this.scaleWidth = width / 2;
                this.scaleHeight = height / 2;
                temp = GraphicsUtils.GetResize(img, scaleWidth, scaleHeight);
                this.image = LImage.CreateImage(scaleWidth, scaleHeight, true);
                this.finalDrawPixels = temp.GetPixels();
                this.nowDrawPixels = (Color[])CollectionUtils.CopyOf(finalDrawPixels);
                if (temp != null)
                {
                    temp.Dispose();
                    temp = null;
                }
            }
            else
            {
                this.scaleWidth = width / 2;
                this.scaleHeight = height / 2;
                temp = GraphicsUtils.GetResize(img, scaleWidth, scaleHeight);
                this.image = LImage.CreateImage(scaleWidth, scaleHeight, true);
                if (back.GetWidth() == scaleWidth
                        && back.GetHeight() == scaleHeight)
                {
                    this.finalBackgroundPixels = back.GetPixels();
                    this.backgroundPixels = (Color[])CollectionUtils
                            .CopyOf(finalBackgroundPixels);
                }
                else
                {
                    LImage tmp = GraphicsUtils.GetResize(back, scaleWidth,
                            scaleHeight);
                    this.finalBackgroundPixels = tmp.GetPixels();
                    if (tmp != null)
                    {
                        tmp.Dispose();
                        tmp = null;
                    }
                    this.backgroundPixels = (Color[])CollectionUtils
                            .CopyOf(finalBackgroundPixels);
                }
                this.finalDrawPixels = temp.GetPixels();
                this.nowDrawPixels = (Color[])CollectionUtils.CopyOf(finalDrawPixels);
            }
            this.SetBlackToWhite(flag);
            if (temp != null)
            {
                temp.Dispose();
                temp = null;
            }
            if (img != null)
            {
                img.Dispose();
                img = null;
            }
            if (back != null)
            {
                back.Dispose();
                back = null;
            }
        }

        public void Reset()
        {
            if (isClose)
            {
                return;
            }
            if (flag)
            {
                pixelCount = min_pixel;
            }
            else
            {
                pixelCount = max_pixel;
            }
            this.visible = true;
            this.nowDrawPixels = (Color[])CollectionUtils.CopyOf(finalDrawPixels);
            this.backgroundPixels = (Color[])CollectionUtils.CopyOf(finalBackgroundPixels);
            this.StartUsePixelThread();
        }

        private static void _Update(int pixelStart, int pixelEnd,
                 Color[] src, Color[] dst, Color[] colors)
        {
            int length = src.Length;
            if (pixelStart < pixelEnd)
            {
                int start = pixelStart + 1;
                int end = pixelEnd + 1;
                if (end > max_pixel)
                {
                    return;
                }
                for (int i = 0; i < length; i++)
                {
                    if (dst[i].PackedValue != LSystem.TRANSPARENT)
                    {
                        for (int pixIndex = start; pixIndex < end; pixIndex++)
                        {
                            if (colors[pixIndex] == src[i])
                            {
                                dst[i].PackedValue = LSystem.TRANSPARENT;
                            }
                            else if (src[i].PackedValue == Color.Black.PackedValue)
                            {
                                dst[i].PackedValue = LSystem.TRANSPARENT;
                            }
                        }
                    }
                }
            }
            else
            {
                int start = pixelEnd - 1;
                int end = pixelStart;
                if (start < 0)
                {
                    return;
                }
                for (int i = 0; i < length; i++)
                {
                    if (dst[i].PackedValue != LSystem.TRANSPARENT)
                    {
                        for (int pixIndex = start; pixIndex < end; pixIndex++)
                        {
                            if (colors[pixIndex] == src[i])
                            {
                                dst[i].PackedValue = LSystem.TRANSPARENT;
                            }
                            else if (src[i].PackedValue == Color.White.PackedValue)
                            {
                                dst[i].PackedValue = LSystem.TRANSPARENT;
                            }
                        }
                    }
                }
            }
        }

        public void CreateUI(GLEx g)
        {
            if (isClose)
            {
                return;
            }
            if (!visible)
            {
                return;
            }
             lock (image)
            {
         
                if (alpha > 0 && alpha < 1)
                {
                    g.SetAlpha(alpha);
                }
                if (!IsComplete() && isDirty)
                {
                    g.DrawTexture2D(image.GetBitmap(), X(), Y(), width, height);
                    isDirty = false;
                }
                else if (!IsComplete())
                {
                    g.DrawTexture2D(image.GetBitmap(), X(), Y(), width, height);
                }
                if (alpha > 0 && alpha < 1)
                {
                    g.SetAlpha(1f);
                }
            }
        }

        private long elapsed;

        public override void Update(long elapsedTime)
        {
            this.elapsed = elapsedTime;
        }

        private class PixelThread : Thread
        {

            PShadowEffect ps;

            public PixelThread(PShadowEffect effect)
            {
                this.ps = effect;
            }

            public override void Run()
            {
                for (; !ps.isClose && !ps.IsComplete(); )
                {
                    if (ps.visible && ps.timer.Action(ps.elapsed))
                    {
                        if (ps.backgroundPixels == null)
                        {
                            if (ps.flag)
                            {
                                _Update(ps.pixelCount, ps.pixelCount += ps.pixelSkip,
                                        ps.finalDrawPixels, ps.nowDrawPixels,
                                        widdershinTrans);
                            }
                            else
                            {
                                _Update(ps.pixelCount, ps.pixelCount -= ps.pixelSkip,
                                        ps.finalDrawPixels, ps.nowDrawPixels, deasilTrans);
                            }
                    
                            ps.image.SetPixels(ps.nowDrawPixels);
                        }
                        else
                        {
                            if (ps.flag)
                            {
                                _Update(ps.pixelCount, ps.pixelCount += ps.pixelSkip,
                                        ps.finalDrawPixels, ps.backgroundPixels,
                                        widdershinTrans);
                            }
                            else
                            {
                                _Update(ps.pixelCount, ps.pixelCount -= ps.pixelSkip,
                                        ps.finalDrawPixels, ps.backgroundPixels,
                                        deasilTrans);
                            }
                            ps.image.SetPixels(ps.backgroundPixels);
                        }
                        ps.isDirty = true;
                    }
                }
            }
        }

        public bool IsVisible()
        {
            return visible;
        }

        public void SetVisible(bool visible)
        {
            this.visible = visible;
        }

        void StartUsePixelThread()
        {
            if (pixelThread == null)
            {
                pixelThread = new PixelThread(this);
                pixelThread.Start();
            }
        }

        void EndUsePixelThread()
        {
            if (pixelThread != null)
            {
                try
                {
                    pixelThread.Interrupt();
                    pixelThread = null;
                }
                catch (Exception)
                {
                    pixelThread = null;
                }
            }
        }

        public bool IsComplete()
        {
            bool stop = flag ? (pixelCount > max_pixel)
                   : (pixelCount < min_pixel);
            if (!stop)
            {
                StartUsePixelThread();
            }
            else
            {
                EndUsePixelThread();
            }
            return stop;
        }

        public void SetDelay(long delay)
        {
            timer.SetDelay(delay);
        }

        public long GetDelay()
        {
            return timer.GetDelay();
        }

        public override int GetHeight()
        {
            return height;
        }

        public override int GetWidth()
        {
            return width;
        }

        public bool IsBlackToWhite()
        {
            return flag;
        }

        public void SetBlackToWhite(bool flag)
        {
            this.flag = flag;
            if (flag)
            {
                pixelCount = min_pixel;
            }
            else
            {
                pixelCount = max_pixel;
            }
        }

        public RectBox GetCollisionBox()
        {
            return GetRect(X(), Y(), width, height);
        }

        public LTexture GetBitmap()
        {
            return image.GetTexture();
        }

        public bool IsClose()
        {
            return isClose;
        }

        public int GetPixelCount()
        {
            return pixelCount;
        }

        public int GetPixelSkip()
        {
            return pixelSkip;
        }

        public void SetPixelSkip(int pixelSkip)
        {
            this.pixelSkip = pixelSkip;
        }

        public void Dispose()
        {
            this.isClose = true;
            this.EndUsePixelThread();
            this.finalDrawPixels = null;
            this.nowDrawPixels = null;
            if (image != null)
            {
                image.Dispose();
                image = null;
            }
        }

    }
}
